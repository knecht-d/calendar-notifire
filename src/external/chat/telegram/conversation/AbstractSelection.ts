import Telegraf, { Context } from "telegraf";
import { TelegrafContext } from "telegraf/typings/context";
import { ExtraReplyMessage } from "telegraf/typings/telegram-types";

export abstract class AbstractSelection<T> {
    constructor(
        protected editMessage: (text: string, extra?: ExtraReplyMessage) => Promise<void>,
        protected messageId: number,
    ) {}
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    static setActions(_bot: Telegraf<TelegrafContext>, _selectionHandler: (ctx: Context, value: string) => void) {}
    abstract requestInput(buildingObject: T): Promise<(selection: string) => T>;
}
