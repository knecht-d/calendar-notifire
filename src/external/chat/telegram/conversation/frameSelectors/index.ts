export { DaySelection } from "./DaySelection";
export { HourSelection } from "./HourSelection";
export { MinuteSelection } from "./MinuteSelection";
export { MonthSelection } from "./MonthSelection";
