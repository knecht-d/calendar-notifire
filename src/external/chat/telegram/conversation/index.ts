export * from "./FrameConfigBuilder";
export * from "./TriggerConfigBuilder";
