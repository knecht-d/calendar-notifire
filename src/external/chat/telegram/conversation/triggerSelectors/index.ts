export { DayOfMonthSelection } from "./DayOfMonthSelection";
export { DayOfWeekSelection } from "./DayOfWeekSelection";
export { HourSelection } from "./HourSelection";
export { MinuteSelection } from "./MinuteSelection";
export { RecurrenceTypeSelection } from "./RecurrenceTypeSelection";
export { ToHourSelection } from "./ToHourSelection";
