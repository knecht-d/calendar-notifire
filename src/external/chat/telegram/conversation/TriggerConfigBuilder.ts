import Telegraf from "telegraf";
import { TelegrafContext } from "telegraf/typings/context";
import { ExtraReplyMessage } from "telegraf/typings/telegram-types";
import { IGenericRecurrence } from "../../../../gateways";
import { ILogger } from "../../../logging";
import { AbstractSelection } from "./AbstractSelection";
import {
    DayOfMonthSelection,
    DayOfWeekSelection,
    HourSelection,
    MinuteSelection,
    RecurrenceTypeSelection,
    ToHourSelection,
} from "./triggerSelectors";
import { handleButtonPress } from "./utils";

export class TriggerConfigBuilder {
    private static ALL_BUILDER: { [messageId: number]: TriggerConfigBuilder } = {};
    private static LOGGER: ILogger;
    config: IGenericRecurrence;
    currentResolver: (selection: string) => void;

    constructor(
        private messageId: number,
        private editMessage: (text: string, extra?: ExtraReplyMessage) => Promise<void>,
    ) {
        this.config = {};
        this.currentResolver = () => {
            return;
        };
        TriggerConfigBuilder.ALL_BUILDER[messageId] = this;
    }

    static init(bot: Telegraf<TelegrafContext>, logger: ILogger) {
        this.LOGGER = logger;
        [
            RecurrenceTypeSelection,
            DayOfMonthSelection,
            DayOfWeekSelection,
            HourSelection,
            ToHourSelection,
            MinuteSelection,
        ].forEach(selection => {
            selection.setActions(bot, this.handleButtonPress.bind(this));
        });
    }

    static handleButtonPress(ctx: TelegrafContext, selection: string) {
        handleButtonPress(this.ALL_BUILDER, this.LOGGER, ctx, selection);
    }

    async requestConfig() {
        await this.requestSelection(RecurrenceTypeSelection);

        if (this.config.type === "m") {
            await this.requestSelection(DayOfMonthSelection);
        } else {
            await this.requestSelection(DayOfWeekSelection);
        }

        await this.requestSelection(HourSelection);

        if (this.config.type === "h") {
            await this.requestSelection(ToHourSelection);
        }

        await this.requestSelection(MinuteSelection);

        delete TriggerConfigBuilder.ALL_BUILDER[this.messageId];

        return this.config;
    }

    getConfig() {
        return this.config;
    }

    private async requestSelection(
        Selection: new (
            editMessage: (text: string, extra?: ExtraReplyMessage) => Promise<void>,
            messageId: number,
        ) => AbstractSelection<IGenericRecurrence>,
    ) {
        const selection = new Selection(this.editMessage, this.messageId);
        const selectionHandler = await selection.requestInput(this.config);
        return new Promise<void>(res => {
            this.currentResolver = selection => {
                this.config = selectionHandler(selection);
                res();
            };
        });
    }
}
