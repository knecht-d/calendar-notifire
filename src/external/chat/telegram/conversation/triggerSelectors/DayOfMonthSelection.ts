import Telegraf, { Context, Extra, Markup } from "telegraf";
import { IGenericRecurrence } from "../../../../../gateways";
import { AbstractSelection } from "../AbstractSelection";

const prefix = "TRIGGER_DAY_OF_MONTH-";
const values = [...Array(31).keys()].map(value => value + 1);

export class DayOfMonthSelection extends AbstractSelection<IGenericRecurrence> {
    static setActions(bot: Telegraf<Context>, selectionHandler: (ctx: Context, value: string) => void): void {
        values.forEach(value => {
            bot.action(`${prefix}${value}`, ctx => selectionHandler(ctx, `${value}`));
        });
    }

    async requestInput(buildingObject: IGenericRecurrence): Promise<(selection: string) => IGenericRecurrence> {
        await this.editMessage(
            `
An welchem Tag im Monat soll die Erinnerung erfolgen?
Gibt es den Tag nicht, dann findet in dem Monat keine Erinnerung statt!
z.B. "30. Februar"`.trim(),
            Extra.HTML().markup(() =>
                Markup.inlineKeyboard(
                    values.map(value => Markup.callbackButton(`${value}`, `${prefix}${value}`)),
                    { columns: 4 },
                ),
            ),
        );
        return (selection: string) => {
            buildingObject.dayOfMonth = Number.parseInt(selection);
            return buildingObject;
        };
    }
}
