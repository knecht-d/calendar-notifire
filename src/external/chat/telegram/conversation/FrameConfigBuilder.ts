import Telegraf from "telegraf";
import { TelegrafContext } from "telegraf/typings/context";
import { ExtraReplyMessage } from "telegraf/typings/telegram-types";
import { IGenericFrame, IGenericRecurrence } from "../../../../gateways";
import { ILogger } from "../../../logging";
import { AbstractSelection } from "./AbstractSelection";
import { DaySelection, HourSelection, MinuteSelection, MonthSelection } from "./frameSelectors";
import { handleButtonPress } from "./utils";

export class FrameConfigBuilder {
    private static ALL_BUILDER: { [messageId: number]: FrameConfigBuilder } = {};
    private static LOGGER: ILogger;
    config: IGenericFrame;
    currentResolver: (selection: string) => void;

    constructor(
        private messageId: number,
        private editMessage: (text: string, extra?: ExtraReplyMessage) => Promise<void>,
    ) {
        this.config = {};
        this.currentResolver = () => {
            return;
        };
        FrameConfigBuilder.ALL_BUILDER[messageId] = this;
    }

    static init(bot: Telegraf<TelegrafContext>, logger: ILogger) {
        this.LOGGER = logger;
        [MonthSelection, DaySelection, MinuteSelection, HourSelection].forEach(selection => {
            selection.setActions(bot, this.handleButtonPress.bind(this));
        });
    }

    static handleButtonPress(ctx: TelegrafContext, selection: string) {
        handleButtonPress(this.ALL_BUILDER, this.LOGGER, ctx, selection);
    }

    async requestConfig(recurrence: IGenericRecurrence) {
        switch (recurrence.type) {
            case "m":
                await this.requestSelection(MonthSelection);
                await this.requestSelection(DaySelection);
                await this.requestSelection(HourSelection);
                break;
            case "d":
                await this.requestSelection(DaySelection);
                await this.requestSelection(HourSelection);
                await this.requestSelection(MinuteSelection);
                break;
            case "h":
                await this.requestSelection(HourSelection);
                await this.requestSelection(MinuteSelection);
                break;
            default:
                throw new Error("Unexpected Recurrence Type");
        }

        delete FrameConfigBuilder.ALL_BUILDER[this.messageId];
        return this.config;
    }

    getConfig() {
        return this.config;
    }

    private async requestSelection(
        Selection: new (
            editMessage: (text: string, extra?: ExtraReplyMessage) => Promise<void>,
            messageId: number,
        ) => AbstractSelection<IGenericFrame>,
    ) {
        const selection = new Selection(this.editMessage, this.messageId);
        const selectionHandler = await selection.requestInput(this.config);
        return new Promise<void>(res => {
            this.currentResolver = selection => {
                this.config = selectionHandler(selection);
                res();
            };
        });
    }
}
