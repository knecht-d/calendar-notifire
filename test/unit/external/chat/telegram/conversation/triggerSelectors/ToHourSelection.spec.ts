import { Extra, Markup } from "telegraf";
import { ToHourSelection } from "../../../../../../../src/external/chat/telegram/conversation/triggerSelectors";

describe("ToHourSelection", () => {
    const values = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23];
    const mockCtx = { mock: "ctx" };
    const MockBot = {
        action: jest.fn((_, handler) => {
            handler(mockCtx);
        }),
    };
    const mockEditMessage = jest.fn();

    beforeEach(() => jest.clearAllMocks());

    describe("setActions", () => {
        it.each([...values])("should set action for %s", value => {
            const mockSelectionHandler = jest.fn();
            ToHourSelection.setActions(MockBot as any, mockSelectionHandler);
            expect(MockBot.action).toHaveBeenCalledWith(`TRIGGER_TO_HOUR-${value}`, expect.any(Function));
        });

        it.each([...values])("should set the correct handler for %s", value => {
            const mockSelectionHandler = jest.fn();
            ToHourSelection.setActions(MockBot as any, mockSelectionHandler);
            expect(mockSelectionHandler).toHaveBeenCalledWith(mockCtx, `${value}`);
        });
    });

    describe("requestInput", () => {
        it("should send the question for other typed", async () => {
            const selection = new ToHourSelection(mockEditMessage, 42);
            await selection.requestInput({});
            expect(mockEditMessage).toHaveBeenCalledWith(
                expect.stringContaining("Bis zu welcher Stunde"),
                expect.any(Object),
            );
        });

        describe("keyboard", () => {
            let selection: ToHourSelection;
            beforeAll(() => {
                jest.spyOn(Extra, "markup").mockImplementation((): any => {
                    return {
                        markup: (cb: () => any) => cb(),
                    };
                });
                jest.spyOn(Markup, "callbackButton").mockImplementation(
                    (desc: string, key: string): any => `${desc}(${key})`,
                );
                jest.spyOn(Markup, "inlineKeyboard");

                selection = new ToHourSelection(mockEditMessage, 42);
            });
            it.each([...values])("should have only values >= %s", async value => {
                await selection.requestInput({ hour: value });

                expect((Markup.inlineKeyboard as jest.Mock).mock.calls).toMatchSnapshot();
            });
        });

        it.each(values)("should set the day on selection of %s", async value => {
            const selection = new ToHourSelection(mockEditMessage, 42);
            const handler = await selection.requestInput({});
            const result = handler(`${value}`);
            expect(result).toEqual({
                hourEnd: value,
            });
        });
    });
});
