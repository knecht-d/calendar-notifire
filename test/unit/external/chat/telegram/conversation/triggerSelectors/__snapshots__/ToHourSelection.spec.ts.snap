// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToHourSelection requestInput keyboard should have only values >= 0 1`] = `
Array [
  Array [
    Array [
      "0(TRIGGER_TO_HOUR-0)",
      "1(TRIGGER_TO_HOUR-1)",
      "2(TRIGGER_TO_HOUR-2)",
      "3(TRIGGER_TO_HOUR-3)",
      "4(TRIGGER_TO_HOUR-4)",
      "5(TRIGGER_TO_HOUR-5)",
      "6(TRIGGER_TO_HOUR-6)",
      "7(TRIGGER_TO_HOUR-7)",
      "8(TRIGGER_TO_HOUR-8)",
      "9(TRIGGER_TO_HOUR-9)",
      "10(TRIGGER_TO_HOUR-10)",
      "11(TRIGGER_TO_HOUR-11)",
      "12(TRIGGER_TO_HOUR-12)",
      "13(TRIGGER_TO_HOUR-13)",
      "14(TRIGGER_TO_HOUR-14)",
      "15(TRIGGER_TO_HOUR-15)",
      "16(TRIGGER_TO_HOUR-16)",
      "17(TRIGGER_TO_HOUR-17)",
      "18(TRIGGER_TO_HOUR-18)",
      "19(TRIGGER_TO_HOUR-19)",
      "20(TRIGGER_TO_HOUR-20)",
      "21(TRIGGER_TO_HOUR-21)",
      "22(TRIGGER_TO_HOUR-22)",
      "23(TRIGGER_TO_HOUR-23)",
    ],
    Object {
      "columns": 4,
    },
  ],
]
`;

exports[`ToHourSelection requestInput keyboard should have only values >= 1 1`] = `
Array [
  Array [
    Array [
      "1(TRIGGER_TO_HOUR-1)",
      "2(TRIGGER_TO_HOUR-2)",
      "3(TRIGGER_TO_HOUR-3)",
      "4(TRIGGER_TO_HOUR-4)",
      "5(TRIGGER_TO_HOUR-5)",
      "6(TRIGGER_TO_HOUR-6)",
      "7(TRIGGER_TO_HOUR-7)",
      "8(TRIGGER_TO_HOUR-8)",
      "9(TRIGGER_TO_HOUR-9)",
      "10(TRIGGER_TO_HOUR-10)",
      "11(TRIGGER_TO_HOUR-11)",
      "12(TRIGGER_TO_HOUR-12)",
      "13(TRIGGER_TO_HOUR-13)",
      "14(TRIGGER_TO_HOUR-14)",
      "15(TRIGGER_TO_HOUR-15)",
      "16(TRIGGER_TO_HOUR-16)",
      "17(TRIGGER_TO_HOUR-17)",
      "18(TRIGGER_TO_HOUR-18)",
      "19(TRIGGER_TO_HOUR-19)",
      "20(TRIGGER_TO_HOUR-20)",
      "21(TRIGGER_TO_HOUR-21)",
      "22(TRIGGER_TO_HOUR-22)",
      "23(TRIGGER_TO_HOUR-23)",
    ],
    Object {
      "columns": 4,
    },
  ],
]
`;

exports[`ToHourSelection requestInput keyboard should have only values >= 2 1`] = `
Array [
  Array [
    Array [
      "2(TRIGGER_TO_HOUR-2)",
      "3(TRIGGER_TO_HOUR-3)",
      "4(TRIGGER_TO_HOUR-4)",
      "5(TRIGGER_TO_HOUR-5)",
      "6(TRIGGER_TO_HOUR-6)",
      "7(TRIGGER_TO_HOUR-7)",
      "8(TRIGGER_TO_HOUR-8)",
      "9(TRIGGER_TO_HOUR-9)",
      "10(TRIGGER_TO_HOUR-10)",
      "11(TRIGGER_TO_HOUR-11)",
      "12(TRIGGER_TO_HOUR-12)",
      "13(TRIGGER_TO_HOUR-13)",
      "14(TRIGGER_TO_HOUR-14)",
      "15(TRIGGER_TO_HOUR-15)",
      "16(TRIGGER_TO_HOUR-16)",
      "17(TRIGGER_TO_HOUR-17)",
      "18(TRIGGER_TO_HOUR-18)",
      "19(TRIGGER_TO_HOUR-19)",
      "20(TRIGGER_TO_HOUR-20)",
      "21(TRIGGER_TO_HOUR-21)",
      "22(TRIGGER_TO_HOUR-22)",
      "23(TRIGGER_TO_HOUR-23)",
    ],
    Object {
      "columns": 4,
    },
  ],
]
`;

exports[`ToHourSelection requestInput keyboard should have only values >= 3 1`] = `
Array [
  Array [
    Array [
      "3(TRIGGER_TO_HOUR-3)",
      "4(TRIGGER_TO_HOUR-4)",
      "5(TRIGGER_TO_HOUR-5)",
      "6(TRIGGER_TO_HOUR-6)",
      "7(TRIGGER_TO_HOUR-7)",
      "8(TRIGGER_TO_HOUR-8)",
      "9(TRIGGER_TO_HOUR-9)",
      "10(TRIGGER_TO_HOUR-10)",
      "11(TRIGGER_TO_HOUR-11)",
      "12(TRIGGER_TO_HOUR-12)",
      "13(TRIGGER_TO_HOUR-13)",
      "14(TRIGGER_TO_HOUR-14)",
      "15(TRIGGER_TO_HOUR-15)",
      "16(TRIGGER_TO_HOUR-16)",
      "17(TRIGGER_TO_HOUR-17)",
      "18(TRIGGER_TO_HOUR-18)",
      "19(TRIGGER_TO_HOUR-19)",
      "20(TRIGGER_TO_HOUR-20)",
      "21(TRIGGER_TO_HOUR-21)",
      "22(TRIGGER_TO_HOUR-22)",
      "23(TRIGGER_TO_HOUR-23)",
    ],
    Object {
      "columns": 4,
    },
  ],
]
`;

exports[`ToHourSelection requestInput keyboard should have only values >= 4 1`] = `
Array [
  Array [
    Array [
      "4(TRIGGER_TO_HOUR-4)",
      "5(TRIGGER_TO_HOUR-5)",
      "6(TRIGGER_TO_HOUR-6)",
      "7(TRIGGER_TO_HOUR-7)",
      "8(TRIGGER_TO_HOUR-8)",
      "9(TRIGGER_TO_HOUR-9)",
      "10(TRIGGER_TO_HOUR-10)",
      "11(TRIGGER_TO_HOUR-11)",
      "12(TRIGGER_TO_HOUR-12)",
      "13(TRIGGER_TO_HOUR-13)",
      "14(TRIGGER_TO_HOUR-14)",
      "15(TRIGGER_TO_HOUR-15)",
      "16(TRIGGER_TO_HOUR-16)",
      "17(TRIGGER_TO_HOUR-17)",
      "18(TRIGGER_TO_HOUR-18)",
      "19(TRIGGER_TO_HOUR-19)",
      "20(TRIGGER_TO_HOUR-20)",
      "21(TRIGGER_TO_HOUR-21)",
      "22(TRIGGER_TO_HOUR-22)",
      "23(TRIGGER_TO_HOUR-23)",
    ],
    Object {
      "columns": 4,
    },
  ],
]
`;

exports[`ToHourSelection requestInput keyboard should have only values >= 5 1`] = `
Array [
  Array [
    Array [
      "5(TRIGGER_TO_HOUR-5)",
      "6(TRIGGER_TO_HOUR-6)",
      "7(TRIGGER_TO_HOUR-7)",
      "8(TRIGGER_TO_HOUR-8)",
      "9(TRIGGER_TO_HOUR-9)",
      "10(TRIGGER_TO_HOUR-10)",
      "11(TRIGGER_TO_HOUR-11)",
      "12(TRIGGER_TO_HOUR-12)",
      "13(TRIGGER_TO_HOUR-13)",
      "14(TRIGGER_TO_HOUR-14)",
      "15(TRIGGER_TO_HOUR-15)",
      "16(TRIGGER_TO_HOUR-16)",
      "17(TRIGGER_TO_HOUR-17)",
      "18(TRIGGER_TO_HOUR-18)",
      "19(TRIGGER_TO_HOUR-19)",
      "20(TRIGGER_TO_HOUR-20)",
      "21(TRIGGER_TO_HOUR-21)",
      "22(TRIGGER_TO_HOUR-22)",
      "23(TRIGGER_TO_HOUR-23)",
    ],
    Object {
      "columns": 4,
    },
  ],
]
`;

exports[`ToHourSelection requestInput keyboard should have only values >= 6 1`] = `
Array [
  Array [
    Array [
      "6(TRIGGER_TO_HOUR-6)",
      "7(TRIGGER_TO_HOUR-7)",
      "8(TRIGGER_TO_HOUR-8)",
      "9(TRIGGER_TO_HOUR-9)",
      "10(TRIGGER_TO_HOUR-10)",
      "11(TRIGGER_TO_HOUR-11)",
      "12(TRIGGER_TO_HOUR-12)",
      "13(TRIGGER_TO_HOUR-13)",
      "14(TRIGGER_TO_HOUR-14)",
      "15(TRIGGER_TO_HOUR-15)",
      "16(TRIGGER_TO_HOUR-16)",
      "17(TRIGGER_TO_HOUR-17)",
      "18(TRIGGER_TO_HOUR-18)",
      "19(TRIGGER_TO_HOUR-19)",
      "20(TRIGGER_TO_HOUR-20)",
      "21(TRIGGER_TO_HOUR-21)",
      "22(TRIGGER_TO_HOUR-22)",
      "23(TRIGGER_TO_HOUR-23)",
    ],
    Object {
      "columns": 4,
    },
  ],
]
`;

exports[`ToHourSelection requestInput keyboard should have only values >= 7 1`] = `
Array [
  Array [
    Array [
      "7(TRIGGER_TO_HOUR-7)",
      "8(TRIGGER_TO_HOUR-8)",
      "9(TRIGGER_TO_HOUR-9)",
      "10(TRIGGER_TO_HOUR-10)",
      "11(TRIGGER_TO_HOUR-11)",
      "12(TRIGGER_TO_HOUR-12)",
      "13(TRIGGER_TO_HOUR-13)",
      "14(TRIGGER_TO_HOUR-14)",
      "15(TRIGGER_TO_HOUR-15)",
      "16(TRIGGER_TO_HOUR-16)",
      "17(TRIGGER_TO_HOUR-17)",
      "18(TRIGGER_TO_HOUR-18)",
      "19(TRIGGER_TO_HOUR-19)",
      "20(TRIGGER_TO_HOUR-20)",
      "21(TRIGGER_TO_HOUR-21)",
      "22(TRIGGER_TO_HOUR-22)",
      "23(TRIGGER_TO_HOUR-23)",
    ],
    Object {
      "columns": 4,
    },
  ],
]
`;

exports[`ToHourSelection requestInput keyboard should have only values >= 8 1`] = `
Array [
  Array [
    Array [
      "8(TRIGGER_TO_HOUR-8)",
      "9(TRIGGER_TO_HOUR-9)",
      "10(TRIGGER_TO_HOUR-10)",
      "11(TRIGGER_TO_HOUR-11)",
      "12(TRIGGER_TO_HOUR-12)",
      "13(TRIGGER_TO_HOUR-13)",
      "14(TRIGGER_TO_HOUR-14)",
      "15(TRIGGER_TO_HOUR-15)",
      "16(TRIGGER_TO_HOUR-16)",
      "17(TRIGGER_TO_HOUR-17)",
      "18(TRIGGER_TO_HOUR-18)",
      "19(TRIGGER_TO_HOUR-19)",
      "20(TRIGGER_TO_HOUR-20)",
      "21(TRIGGER_TO_HOUR-21)",
      "22(TRIGGER_TO_HOUR-22)",
      "23(TRIGGER_TO_HOUR-23)",
    ],
    Object {
      "columns": 4,
    },
  ],
]
`;

exports[`ToHourSelection requestInput keyboard should have only values >= 9 1`] = `
Array [
  Array [
    Array [
      "9(TRIGGER_TO_HOUR-9)",
      "10(TRIGGER_TO_HOUR-10)",
      "11(TRIGGER_TO_HOUR-11)",
      "12(TRIGGER_TO_HOUR-12)",
      "13(TRIGGER_TO_HOUR-13)",
      "14(TRIGGER_TO_HOUR-14)",
      "15(TRIGGER_TO_HOUR-15)",
      "16(TRIGGER_TO_HOUR-16)",
      "17(TRIGGER_TO_HOUR-17)",
      "18(TRIGGER_TO_HOUR-18)",
      "19(TRIGGER_TO_HOUR-19)",
      "20(TRIGGER_TO_HOUR-20)",
      "21(TRIGGER_TO_HOUR-21)",
      "22(TRIGGER_TO_HOUR-22)",
      "23(TRIGGER_TO_HOUR-23)",
    ],
    Object {
      "columns": 4,
    },
  ],
]
`;

exports[`ToHourSelection requestInput keyboard should have only values >= 10 1`] = `
Array [
  Array [
    Array [
      "10(TRIGGER_TO_HOUR-10)",
      "11(TRIGGER_TO_HOUR-11)",
      "12(TRIGGER_TO_HOUR-12)",
      "13(TRIGGER_TO_HOUR-13)",
      "14(TRIGGER_TO_HOUR-14)",
      "15(TRIGGER_TO_HOUR-15)",
      "16(TRIGGER_TO_HOUR-16)",
      "17(TRIGGER_TO_HOUR-17)",
      "18(TRIGGER_TO_HOUR-18)",
      "19(TRIGGER_TO_HOUR-19)",
      "20(TRIGGER_TO_HOUR-20)",
      "21(TRIGGER_TO_HOUR-21)",
      "22(TRIGGER_TO_HOUR-22)",
      "23(TRIGGER_TO_HOUR-23)",
    ],
    Object {
      "columns": 4,
    },
  ],
]
`;

exports[`ToHourSelection requestInput keyboard should have only values >= 11 1`] = `
Array [
  Array [
    Array [
      "11(TRIGGER_TO_HOUR-11)",
      "12(TRIGGER_TO_HOUR-12)",
      "13(TRIGGER_TO_HOUR-13)",
      "14(TRIGGER_TO_HOUR-14)",
      "15(TRIGGER_TO_HOUR-15)",
      "16(TRIGGER_TO_HOUR-16)",
      "17(TRIGGER_TO_HOUR-17)",
      "18(TRIGGER_TO_HOUR-18)",
      "19(TRIGGER_TO_HOUR-19)",
      "20(TRIGGER_TO_HOUR-20)",
      "21(TRIGGER_TO_HOUR-21)",
      "22(TRIGGER_TO_HOUR-22)",
      "23(TRIGGER_TO_HOUR-23)",
    ],
    Object {
      "columns": 4,
    },
  ],
]
`;

exports[`ToHourSelection requestInput keyboard should have only values >= 12 1`] = `
Array [
  Array [
    Array [
      "12(TRIGGER_TO_HOUR-12)",
      "13(TRIGGER_TO_HOUR-13)",
      "14(TRIGGER_TO_HOUR-14)",
      "15(TRIGGER_TO_HOUR-15)",
      "16(TRIGGER_TO_HOUR-16)",
      "17(TRIGGER_TO_HOUR-17)",
      "18(TRIGGER_TO_HOUR-18)",
      "19(TRIGGER_TO_HOUR-19)",
      "20(TRIGGER_TO_HOUR-20)",
      "21(TRIGGER_TO_HOUR-21)",
      "22(TRIGGER_TO_HOUR-22)",
      "23(TRIGGER_TO_HOUR-23)",
    ],
    Object {
      "columns": 4,
    },
  ],
]
`;

exports[`ToHourSelection requestInput keyboard should have only values >= 13 1`] = `
Array [
  Array [
    Array [
      "13(TRIGGER_TO_HOUR-13)",
      "14(TRIGGER_TO_HOUR-14)",
      "15(TRIGGER_TO_HOUR-15)",
      "16(TRIGGER_TO_HOUR-16)",
      "17(TRIGGER_TO_HOUR-17)",
      "18(TRIGGER_TO_HOUR-18)",
      "19(TRIGGER_TO_HOUR-19)",
      "20(TRIGGER_TO_HOUR-20)",
      "21(TRIGGER_TO_HOUR-21)",
      "22(TRIGGER_TO_HOUR-22)",
      "23(TRIGGER_TO_HOUR-23)",
    ],
    Object {
      "columns": 4,
    },
  ],
]
`;

exports[`ToHourSelection requestInput keyboard should have only values >= 14 1`] = `
Array [
  Array [
    Array [
      "14(TRIGGER_TO_HOUR-14)",
      "15(TRIGGER_TO_HOUR-15)",
      "16(TRIGGER_TO_HOUR-16)",
      "17(TRIGGER_TO_HOUR-17)",
      "18(TRIGGER_TO_HOUR-18)",
      "19(TRIGGER_TO_HOUR-19)",
      "20(TRIGGER_TO_HOUR-20)",
      "21(TRIGGER_TO_HOUR-21)",
      "22(TRIGGER_TO_HOUR-22)",
      "23(TRIGGER_TO_HOUR-23)",
    ],
    Object {
      "columns": 4,
    },
  ],
]
`;

exports[`ToHourSelection requestInput keyboard should have only values >= 15 1`] = `
Array [
  Array [
    Array [
      "15(TRIGGER_TO_HOUR-15)",
      "16(TRIGGER_TO_HOUR-16)",
      "17(TRIGGER_TO_HOUR-17)",
      "18(TRIGGER_TO_HOUR-18)",
      "19(TRIGGER_TO_HOUR-19)",
      "20(TRIGGER_TO_HOUR-20)",
      "21(TRIGGER_TO_HOUR-21)",
      "22(TRIGGER_TO_HOUR-22)",
      "23(TRIGGER_TO_HOUR-23)",
    ],
    Object {
      "columns": 4,
    },
  ],
]
`;

exports[`ToHourSelection requestInput keyboard should have only values >= 16 1`] = `
Array [
  Array [
    Array [
      "16(TRIGGER_TO_HOUR-16)",
      "17(TRIGGER_TO_HOUR-17)",
      "18(TRIGGER_TO_HOUR-18)",
      "19(TRIGGER_TO_HOUR-19)",
      "20(TRIGGER_TO_HOUR-20)",
      "21(TRIGGER_TO_HOUR-21)",
      "22(TRIGGER_TO_HOUR-22)",
      "23(TRIGGER_TO_HOUR-23)",
    ],
    Object {
      "columns": 4,
    },
  ],
]
`;

exports[`ToHourSelection requestInput keyboard should have only values >= 17 1`] = `
Array [
  Array [
    Array [
      "17(TRIGGER_TO_HOUR-17)",
      "18(TRIGGER_TO_HOUR-18)",
      "19(TRIGGER_TO_HOUR-19)",
      "20(TRIGGER_TO_HOUR-20)",
      "21(TRIGGER_TO_HOUR-21)",
      "22(TRIGGER_TO_HOUR-22)",
      "23(TRIGGER_TO_HOUR-23)",
    ],
    Object {
      "columns": 4,
    },
  ],
]
`;

exports[`ToHourSelection requestInput keyboard should have only values >= 18 1`] = `
Array [
  Array [
    Array [
      "18(TRIGGER_TO_HOUR-18)",
      "19(TRIGGER_TO_HOUR-19)",
      "20(TRIGGER_TO_HOUR-20)",
      "21(TRIGGER_TO_HOUR-21)",
      "22(TRIGGER_TO_HOUR-22)",
      "23(TRIGGER_TO_HOUR-23)",
    ],
    Object {
      "columns": 4,
    },
  ],
]
`;

exports[`ToHourSelection requestInput keyboard should have only values >= 19 1`] = `
Array [
  Array [
    Array [
      "19(TRIGGER_TO_HOUR-19)",
      "20(TRIGGER_TO_HOUR-20)",
      "21(TRIGGER_TO_HOUR-21)",
      "22(TRIGGER_TO_HOUR-22)",
      "23(TRIGGER_TO_HOUR-23)",
    ],
    Object {
      "columns": 4,
    },
  ],
]
`;

exports[`ToHourSelection requestInput keyboard should have only values >= 20 1`] = `
Array [
  Array [
    Array [
      "20(TRIGGER_TO_HOUR-20)",
      "21(TRIGGER_TO_HOUR-21)",
      "22(TRIGGER_TO_HOUR-22)",
      "23(TRIGGER_TO_HOUR-23)",
    ],
    Object {
      "columns": 4,
    },
  ],
]
`;

exports[`ToHourSelection requestInput keyboard should have only values >= 21 1`] = `
Array [
  Array [
    Array [
      "21(TRIGGER_TO_HOUR-21)",
      "22(TRIGGER_TO_HOUR-22)",
      "23(TRIGGER_TO_HOUR-23)",
    ],
    Object {
      "columns": 4,
    },
  ],
]
`;

exports[`ToHourSelection requestInput keyboard should have only values >= 22 1`] = `
Array [
  Array [
    Array [
      "22(TRIGGER_TO_HOUR-22)",
      "23(TRIGGER_TO_HOUR-23)",
    ],
    Object {
      "columns": 4,
    },
  ],
]
`;

exports[`ToHourSelection requestInput keyboard should have only values >= 23 1`] = `
Array [
  Array [
    Array [
      "23(TRIGGER_TO_HOUR-23)",
    ],
    Object {
      "columns": 4,
    },
  ],
]
`;
